q2cli/core/usage.py

```diff
@@ -0,0 +1,156 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from qiime2.core.type.primitive import Bool
+
+import qiime2.sdk.usage as usage
+from qiime2 import Metadata, MetadataColumn
+from qiime2.sdk.util import (
+    is_metadata_type,
+    is_metadata_column_type,
+    is_visualization_type,
+)
+
+from q2cli.util import to_cli_name
+from q2cli.util import to_snake_case
+
+
+class CLIUsage(usage.Usage):
+    def __init__(self):
+        super().__init__()
+        self._recorder = []
+        self._init_data_refs = dict()
+
+    def _init_data_(self, ref, factory):
+        self._init_data_refs[ref] = factory
+        data = factory()
```

Comment:
This could potentially be expensive (especially for poorly-behaved usage examples). I think it would be fine to not do these type checks, and instead below in `_template_metadata` you can just invert the logic there - lean on the `signature` (the signature can tell you which of your input opts _should_ be metadata). 

Ultimately the typechecking will be handled by the framework, when a usage example is exercised with the `ExecutionUsage` driver.

Reply:
I was trying to ensure metadata column names are accessible.  When a metadata column is passed in as a factory, e.g.,

```python
def identity_with_metadata_column_from_factory(use):
    ints = use.init_data('ints', ints1_factory)
    mdc = use.init_data('mdc', mdc1_factory)


    use.action(
        UsageAction(plugin_id='dummy_plugin',
                    action_id='identity_with_metadata_column'),
        UsageInputs(ints=ints, metadata=mdc),
        UsageOutputNames(out='out'),
    )
```
I think the only way to get the column name is to call the factory.

Perhaps the solution is to only call a factory when it's a metadata column factory.

Reply:
Gotcha - the `ref` in the `_init_data_` signature is the MDC's name, so you can just use the ref as-is. Another reason for us to get docstrings into the usage API!

Reply:
So in the case above, your saying the MDC's name for `mdc1_factory` is `'mdc'`?

I thought that the MDC's name for the `mdc1_factory` was `'a'`.  Is that not what we're after?  
 
```python
def mdc1_factory():
    return CategoricalMetadataColumn(pd.Series(['1', '2', '3'],
                                     name='a',
                                     index=pd.Index(['0', '1', '2'],
                                                    name='id')))
```

I think getting the actual name of the column is only possible after calling the factory.  Am I wrong about that?  I feel like I might be misunderstanding something here.

Reply:
No worries, I misspoke, I was thinking about `get_metadata_column`, but was talking about `_init_data_` (`get_metadata_column` includes a ref to the column name for just this case). I think this might be a good reason to not support the case you pointed out above (`identity_with_metadata_column_from_factory`). I recall discussing this with @ebolyen, but I don't recall any of the specifics, but anyway, I think that maybe this is an indication that maybe the API needs to be tweaked a bit.

Reply:
BTW, @ebolyen, I think when we chatted about this before, I referred to it as "the least common denominator" issue - since we _must_ have a metadata file for interfaces like q2cli, maybe that means that Usage API needs to follow a similar guideline of working against the "least common denominator" (which is `Metadata` only, not `Metadata` and `MetadataColumn`).

@andrewsanchez - what do you think of this?

Reply:
What if we change the signature of `init_data` to something like this:

```diff
modified   qiime2/core/testing/examples.py
@@ -193,7 +193,7 @@ def identity_with_metadata_column_get_mdc(use):
 
 def identity_with_metadata_column_from_factory(use):
     ints = use.init_data('ints', ints1_factory)
-    mdc = use.init_data('mdc', mdc1_factory)
+    mdc = use.init_data('mdc', mdc1_factory, name='a')
 
     use.action(
         UsageAction(plugin_id='dummy_plugin',
modified   qiime2/sdk/usage.py
@@ -217,8 +217,8 @@ class Usage(metaclass=abc.ABCMeta):
     def __init__(self):
         self._scope = Scope()
 
-    def init_data(self, ref, factory):
-        value = self._init_data_(ref, factory)
+    def init_data(self, ref, factory, name=None):
+        value = self._init_data_(ref, factory, name)
```

Then CLIUsage can do something like this:
```diff
modified   q2cli/core/usage.py
@@ -28,8 +28,10 @@ class CLIUsage(usage.Usage):
         self._recorder = []
         self._init_data_refs = dict()
 
-    def _init_data_(self, ref, factory):
+    def _init_data_(self, ref, factory, name=None):
         self._init_data_refs[ref] = factory
+        if name:
+            return name, ref
         return ref
@@ -144,20 +146,15 @@ class CLIUsage(usage.Usage):
                     for k, v in data.items():
                         mds_t.append(f"{file_param} {k}.tsv")
             elif is_metadata_column_type(qtype):
-                try:
-                    col, md = input_opts[i]
-                except ValueError:
-                    # The metadata column was passed in as a factory. The
-                    # factory needs to be called so we can get the column name
-                    md = input_opts[i]
-                    data = self._init_data_refs[md]()
-                    col = data.name
+                col, md = input_opts[i]
                 mds_t.append(f"{file_param} {md}.tsv")
                 mds_t.append(f"{col_param} '{col}'")
         return mds_t

```

Reply:
That's an interesting idea - do you think that would be confusing for the other types of data, like Artifacts? I'm kinda thinking that maybe we should refactor `init_data` into two or three new methods: `init_artifact`, `init_metadata`, and `init_metadata_column`. The factories would return data to be fed into the `Artifact`, `Metadata`, and `MetadataColumn` constructors, rather than instances of the aforementioned data types. Then, the scope can actually keep track of the types of the data, which then would let the interface drivers do something specific with that info. For example, with this `identity_with_metadata_column_from_factory` case, the `CLIUsage` driver could actually just "materialize" a new `Metadata` tsv file with one column, the column used and referenced in the `init_metadata_column` dataset.

Reply:
> That's an interesting idea - do you think that would be confusing for the other types of data, like Artifacts? 

With sufficient documentation, I don't think so.   I think my proposal is a little hacky, but perhaps not terribly so?  Your proposal definitely seems better but maybe more of a "nice to have," worth implementing to improve the API later?  Since my proposal is such a small, non-breaking* change, perhaps we can use it as an interim solution and open an issue for an API improvement later.

It would be non-breaking with the addition of something like:

```diff
+import inspect
 import re
 import types
 
@@ -217,11 +218,14 @@ class Usage(metaclass=abc.ABCMeta):
     def __init__(self):
         self._scope = Scope()
 
-    def init_data(self, ref, factory):
-        value = self._init_data_(ref, factory)
+    def init_data(self, ref, factory, name=None):
+        if "name" in inspect.getargspec(self._init_data_).args:
+            value = self._init_data_(ref, factory, name)
+        else:
+            value = self._init_data_(ref, factory)
         return self._push_record(ref, value)
 
-    def _init_data_(self, ref, factory):
+    def _init_data_(self, ref, factory, name=None):
         raise NotImplementedError
 
     def merge_metadata(self, ref, *records):
```

Reply:
> With sufficient documentation, I don't think so

Interesting! I'm curious what you have in mind. The case that I'm wondering about can be summarized like this:

```python
def ints1_factory():
    return Artifact.import_data(IntSequence1, [0, 1, 2])

def concatenate_ints_simple(use):
    ints_a = use.init_data('ints_a', ints1_factory, name='spleen')
   ...
```

What does the name mean in this case? How does it differ from the `ints_a` name (ref) that is being specified?

> but maybe more of a "nice to have," worth implementing to improve the API later?

We aren't in a huge rush, I don't mind ironing out the API a bit more. When we merged Artifact API into 2020.2 we recognized that the API wasn't quite complete, and we would deal with issues as they cropped up.